docs/doc_gen_test.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/python
+#
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for the `doc_gen` module."""
+
+import unittest
+import doc_gen
+
+
+class ProcessFileTest(unittest.TestCase):
+
+    def setUp(self):
+        self.docgen_readfile = None
+
+
+    def tearDown(self):
+        if self.docgen_readfile is not None:
+            doc_gen.ReadFileRaw = self.docgen_readfile
+
+
+    def _setCustomReadFile(self, filepath, contents):
+        self.docgen_readfile = doc_gen.ReadFileRaw
+        def readFileCustom(filename):
+            if filename == filepath:
+                return contents
+            else:
+                raise Exception('Expected file: "%s", received: "%s"' %
+                        (filepath, filename))
+
+        doc_gen.ReadFileRaw = readFileCustom
+
+
+    def testPreC99Comments(self):
+        source_file = 'file.c'
+        source_file_fullpath = '../src/examples/%s' % source_file
```

Comment:
Would it make sense to base paths on the path of the test script?

e.g

script_path = os.path.realpath(os.path.dirname(sys.argv[0]))
source_file = os.path.join([script_path, "../src/examples", source_file])

Reply:
The paths are actually synthetic, because the test mocks out the method `ReadFileRaw()`, so these paths do not need to exist — in fact, they will not be read from the file system, and will only process the contents of the variable `source_file_contents` in the same test.

I've renamed the path from `../src/examples` to `/source/code/` to demonstrate this, and the tests still pass.